src/database/database.ts

```diff
@@ -0,0 +1,268 @@
+      method: 'GET',
+      url: this.dbUrl,
+    };
+    return this.httpClient.send(req)
+      .then((resp) => {
+        if (!resp.text) {
+          throw new FirebaseAppError(AppErrorCodes.INTERNAL_ERROR, 'HTTP response missing data.');
+        }
+        return resp.text;
+      })
+      .catch((err) => {
+        throw this.handleError(err);
+      });
+  }
+
+  /**
+   * Gets the currently applied security rules as a parsed JSON object. Any comments in
+   * the original source are stripped away.
+   *
+   * @return {Promise<object>} A promise fulfilled with the parsed rules source.
+   */
+  public getRulesJSON(): Promise<object> {
+    const req: HttpRequestConfig = {
+      method: 'GET',
+      url: this.dbUrl,
+      data: {format: 'strict'},
+    };
+    return this.httpClient.send(req)
+      .then((resp) => {
+        return resp.data;
+      })
+      .catch((err) => {
+        throw this.handleError(err);
+      });
+  }
+
+  /**
+   * Sets the specified rules on the Firebase Database instance. If the rules source is
+   * specified as a string or a Buffer, it may include comments.
+   *
+   * @param {string|Buffer|object} source Source of the rules to apply. Must not be `null`
+   *  or empty.
+   * @return {Promise<void>} Resolves when the rules are set on the Database.
+   */
+  public setRules(source: string | Buffer | object): Promise<void> {
+    if (!validator.isNonEmptyString(source) &&
+      !validator.isBuffer(source)  &&
+      !validator.isNonNullObject(source)) {
+      const error = new FirebaseDatabaseError({
+        code: 'invalid-argument',
+        message: 'Source must be a non-empty string, Buffer or an object.',
+      });
+      return Promise.reject(error);
+    }
+
+    const req: HttpRequestConfig = {
+      method: 'PUT',
+      url: this.dbUrl,
+      data: source,
+      headers: {
+        'content-type': 'application/json; charset=utf-8',
+      },
+    };
+    return this.httpClient.send(req)
+      .then(() => {
+        return;
+      })
+      .catch((err) => {
+        throw this.handleError(err);
+      });
+  }
+
+  private handleError(err: Error): Error {
+    if (err instanceof HttpError) {
+      return new FirebaseDatabaseError({
+        code: AppErrorCodes.INTERNAL_ERROR,
+        message: this.getErrorMessage(err),
+      });
+    }
+    return err;
+  }
+
+  private getErrorMessage(err: HttpError): string {
+    const intro = 'Error while accessing security rules';
+    try {
+      const body: {error?: string} = err.response.data;
+      if (body && body.error) {
+        return `${intro}: ${body.error.trim()}`;
+      }
+    } catch {
+      // Ignore parsing errors
+    }
+
+    return `${intro}: ${err.response.text}`;
+  }
+}
+
+// HACKY, found in the database.d.ts
+export type EventType = 'value' | 'child_added' | 'child_changed' | 'child_moved' | 'child_removed';
+
+
+// TODO: Fix hacky ?? 
+export function database(app?: FirebaseApp, url?: string): Database {
```

Comment:
This is obviously bad design because TypeScript doesn't have kwargs like Python. 

Apart from making one of the parameters mandatory (make `FirebaseApp` last?), the parameter could also be an object that has an interface dictating the different parameters (which would be more scalable too to future additions). If this route is taken, I think it'd be better for `auth(...)` to possibly use it as well and make it be consistent across all the top-level service functions.

Reply:
May be expose 2 functions?

```
database(app?: FirebaseApp): Database
databaseWithURL(url: string, app?: FirebaseApp): Database
```

Reply:
I like that idea; it's consistent with the service function for Auth while having a method allowing for a url string.

I gave it some thought and this seems best. Do you think there's the possibility of adding more optional arguments beyond app/url? Then possibly having an object encapsulate the parameters would be nice (it's more future-proof?) and have the fields as optional.